### PR TITLE
feat(events): Add payment reject audit events

### LINKED ISSUE #4669: Payment Reject

### PR DESCRIPTION
**Pull request by AnishKacham**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [X] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Pass along request_state to payment_core
Modify the UpdateTracker trait to accept request state
Modify the PaymentReject implementation of UpdateTracker to generate an event

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

Files changed
1. `crates/router/src/core/payments/operations/payment_reject.rs`
2. `crates/router/src/events/audit_events.rs`
## Motivation and Context

Resolves #4669 

## How did you test it?

This requires FRM related flows to be tested, and may not be tested locally.

## Checklist

- [X] I formatted the code `cargo +nightly fmt --all`
- [X] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

**Comment by tsdk02** [comment]
Hey @AnishKacham ,
Can you share the event json or a screenshot of the event,
Can refer to pr https://github.com/juspay/hyperswitch/pull/4166 for reference.

For the UI in the screenshot, you can find the docs to set up Kafka, kafka-ui (the interface in the screenshot) and ClickHouse using Docker Compose here: [https://github.com/juspay/hyperswitch/tree/main/crates/analytics/docs](https://github.com/juspay/hyperswitch/tree/main/crates/analytics/docs)

and then try generating said events and monitor logs in kafka-ui.

Also the docs are self-explanatory but still if in any doubt ping us in [this](https://hyperswitch-io.slack.com/archives/C04JXRZ338A) slack channel

File: crates/router/src/core/payments/operations/payment_reject.rs [hunk 1]
```diff
@@ -264,7 +265,13 @@ impl<F: Clone> UpdateTracker<F, PaymentData<F>, PaymentsCancelRequest> for Payme
             )
             .await
             .to_not_found_response(errors::ApiErrorResponse::PaymentNotFound)?;
-
+        
+        req_state
+            .event_context
+            .event(AuditEvent::new(AuditEventType::PaymentReject))
+            .with(payment_data.to_event())
+            .emit();
```
Comment (tsdk02) [inline crates/router/src/core/payments/operations/payment_reject.rs:273]: Can you send `error_code` and `error_message` as well to be logged here
You can take reference from this PR: [https://github.com/juspay/hyperswitch/pull/4913/files](https://github.com/juspay/hyperswitch/pull/4913/files)

File: crates/router/src/events/audit_events.rs [hunk 2]
```diff
@@ -27,6 +27,7 @@ pub enum AuditEventType {
         capture_amount: Option<MinorUnit>,
         multiple_capture_count: Option<i16>,
     },
+    PaymentReject,
```
Comment (tsdk02) [inline crates/router/src/events/audit_events.rs:30]: Add `error_code` and `error_message` here as well.

**Comment by tsdk02** [comment]
Hey @AnishKacham, I have confirmed with the concerned team, and `PaymentReject` is an internal operation and it cannot be tested independently.
Please make the above changes and your PR will be good to go.

**Comment by AnishKacham** [comment]
> Hey @AnishKacham, I have confirmed with the concerned team, and `PaymentReject` is an internal operation and it cannot be tested independently. Please make the above changes and your PR will be good to go.

I'll make the changes

**Comment by tsdk02** [comment]
Hey @AnishKacham, a gentle reminder to add the changes to this PR, so it can be merged

**Comment by AnishKacham** [comment]
> Hey @AnishKacham, a gentle reminder to add the changes to this PR, so it can be merged

Done. Let me know if something needs to be addressed.

File: crates/router/src/core/payments/operations/payment_reject.rs [hunk 3]
```diff
@@ -264,6 +265,16 @@ impl<F: Clone> UpdateTracker<F, PaymentData<F>, PaymentsCancelRequest> for Payme
             )
             .await
             .to_not_found_response(errors::ApiErrorResponse::PaymentNotFound)?;
+        let error_code = payment_data.payment_attempt.error_code;
```
Comment (tsdk02) [inline crates/router/src/core/payments/operations/payment_reject.rs:268]:
```suggestion
        let error_code = payment_data.payment_attempt.error_code.clone();
```
Consider cloning it.

File: crates/router/src/core/payments/operations/payment_reject.rs [hunk 4]
```diff
@@ -264,6 +265,16 @@ impl<F: Clone> UpdateTracker<F, PaymentData<F>, PaymentsCancelRequest> for Payme
...
+        let error_message = payment_data.payment_attempt.error_message;
```
Comment (tsdk02) [inline crates/router/src/core/payments/operations/payment_reject.rs:269]:
```suggestion
        let error_message = payment_data.payment_attempt.error_message.clone();
```

Here as well. CI checks are failing.

**Review by tsdk02 — APPROVED** [review]

**Review by Chethan-rao — DISMISSED** [review]

**Review by Chethan-rao — APPROVED** [review]

**Review by tsdk02 — APPROVED** [review]

**Pull request merged**